toolchain/check/node_stack.h

```diff
@@ -194,6 +208,38 @@ class NodeStack {
                    << "; see value in ParseNodeKindToIdKind";
   }
 
+  // Pops the top of the stack and returns the parse_node and the ID.
+  template <Parse::NodeCategory RequiredParseCategory>
+  auto PopWithParseNode() -> auto {
+    constexpr std::optional<IdKind> RequiredIdKind =
+        ParseNodeCategoryToIdKind(RequiredParseCategory);
+    CARBON_CHECK(RequiredIdKind.has_value())
+        << "NodeCategory " << RequiredParseCategory
+        << " doesn't correspond to an IdKind";
+    auto node_id_cast = [&](auto back) {
+      using NodeIdT = Parse::NodeIdInCategory<RequiredParseCategory>;
+      return std::pair<NodeIdT, decltype(back.second)>(back);
+    };
+    if constexpr (RequiredIdKind == IdKind::InstId) {
+      auto back = PopWithParseNode<SemIR::InstId>();
+      RequireParseCategory<RequiredParseCategory>(back.first);
+      return node_id_cast(back);
+    }
+    if constexpr (RequiredIdKind == IdKind::InstBlockId) {
+      auto back = PopWithParseNode<SemIR::InstBlockId>();
+      RequireParseCategory<RequiredParseCategory>(back.first);
+      return node_id_cast(back);
+    }
+    if constexpr (RequiredIdKind == IdKind::NameId) {
+      auto back = PopWithParseNode<SemIR::NameId>();
+      RequireParseCategory<RequiredParseCategory>(back.first);
+      return node_id_cast(back);
+    }
+    CARBON_FATAL() << "Unpoppable IdKind for parse category: "
+                   << RequiredParseCategory
+                   << "; see value in ParseNodeCategoryToIdKind";
```

Comment:
`static_assert` here too would be good I think. (We'll need to change most of the `if constexpr`s to `else if constexpr`s and put this in a final `else`.)

Reply:
How do you write that? I tried
```
if constexpr (...) {
} else if constexpr (...) {
} else {
  static_assert(false);
}
```
and it gave me a compile error.

Reply:
It looks like we don't require a new enough Clang for that: https://clang.llvm.org/cxx_dr_status.html#2518

In the mean time we could do this:
```c++
if constexpr (cond1) {
  // ...
} else if constexpr (cond2) {
  // ...
// instead of `} else if constexpr (cond3) {`:
} else {
  static_assert(cond3);
  // ...
}
```